### PR TITLE
Ability Scaling

### PR DESCRIPTION
**Pull request by Valo56**
Scaling to damages and healing for Devices and Force abilities. Extensive rework of cast times, cooldowns and effects. Some abilities (Frag, Spark, Disturbance) can be spammed almost every round, others are meant to be used two or three times per minute. Per target damages based on and comparable to expected damage output of basic attacks (which is roughly 110 DMG at max level with max stat investment and 60% hit chance). Minor VFX changes to some abilities (Lightning, Disturbance, Spark, Burst) to make them more flashy. If a character is attacking as cast times (which are all between 1 and 3 seconds now among affected abilities) finish then the ability fails.

A full list of changes will be made available for inclusion to the design doc after testing.

Also, credit to Morrigan/Zaela as much of this is based on work they initially put in on this project.

### CODE REVIEW

File: SWLOR.Game.Server/Service/AbilityService/RecastGroup.cs [hunk 1]
```diff
@@ -25,14 +25,14 @@ public enum RecastGroup
         MindTrick = 8,
         [RecastGroup("Force Burst", "Force Burst", true)]
         ForceBurst = 9,
-        [RecastGroup("Force Body", "Force Body", true)]
-        ForceBody = 10,
+        [RecastGroup("Force Restore", "Force Restore", true)]
+        ForceRestore = 10,
```
Comment (zunath) [inline SWLOR.Game.Server/Service/AbilityService/RecastGroup.cs:29]: Changing the enumeration name on anything which is already stored in the database will cause errors when those players log into the server. You need to write a migration to do this rename on the raw player JSON files. In this instance, I would also be ok with just deleting the recast timer dictionaries on player records.

File: SWLOR.Game.Server/Service/AbilityService/RecastGroup.cs [hunk 2]
```diff
@@ -25,14 +25,14 @@ public enum RecastGroup
...
         [RecastGroup("Force Drain", "Force Drain", true)]
         ForceDrain = 11,
         [RecastGroup("Force Lightning", "F. Lightning", true)]
         ForceLightning = 12,
-        [RecastGroup("Force Mind", "Force Mind", true)]
-        ForceMind = 13,
+        [RecastGroup("Grenades", "Grenades", true)]
+        Grenades = 13,
```
Comment (zunath) [inline SWLOR.Game.Server/Service/AbilityService/RecastGroup.cs:35]: I don't understand this change. If Force Mind was removed, fine, but all of the grenades have their own recast timers now. Please check and verify, please explain to me if I'm off base.

File: SWLOR.Game.Server/Feature/UsePerkFeat.cs [hunk 3]
```diff
@@ -239,8 +239,12 @@ void CompleteActivation(string id, float abilityRecastDelay)
                 DeleteLocalInt(activator, id);
 
                 // Moved during casting or activator died. Cancel the activation.
-                if (GetLocalInt(activator, id) == (int)ActivationStatus.Interrupted || GetCurrentHitPoints(activator) <= 0)
+                if (GetLocalInt(activator, id) == (int)ActivationStatus.Interrupted || GetCurrentHitPoints(activator) <= 0 || GetCurrentAction(activator) != ActionType.Invalid)
+                {
+                    Activity.ClearBusy(activator);
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/UsePerkFeat.cs:244]: I don't know if this is possible or not, but clearing the busy state here can affect more than just casting. Can you confirm there's not a way to cancel the busy state on any non-casting actions - for example, harvesting?

File: SWLOR.Game.Server/Feature/StatusEffectDefinition/ForceRestoreStatusEffectDefinition.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        }
+        private void ForceBody2(StatusEffectBuilder builder)
+        {
+            builder.Create(StatusEffectType.ForceBody2)
+                .Name("Force Body II")
+                .EffectIcon(EffectIconType.TemporaryHitpoints)
+                .CannotReplace(StatusEffectType.ForceBody2)
+                .GrantAction((source, target, length, data) =>
+                {
+                    var willBonus = GetAbilityModifier(AbilityType.Willpower, source);
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/StatusEffectDefinition/ForceRestoreStatusEffectDefinition.cs:53]: The logic here is done in two places. Please wrap it in a method and call that method so we don't have to make edits in more than one location.

File: SWLOR.Game.Server/Feature/StatusEffectDefinition/ForceRestoreStatusEffectDefinition.cs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        {
+            builder.Create(StatusEffectType.ForceMind1)
+                .Name("Force Mind I")
+                .EffectIcon(EffectIconType.TemporaryHitpoints)
+                .CannotReplace(StatusEffectType.ForceMind1)
+                .GrantAction((source, target, length, data) =>
+                {
+                    var willBonus = GetAbilityModifier(AbilityType.Willpower, source);
+                    var willRestore = willBonus / 2;
+                    Stat.RestoreFP(target, willRestore);
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/StatusEffectDefinition/ForceRestoreStatusEffectDefinition.cs:77]: Did you mean to do STM restore here?

File: SWLOR.Game.Server/Service/AbilityService/RecastGroup.cs (hunk 2 above)
Comment (Valo56) [reply SWLOR.Game.Server/Service/AbilityService/RecastGroup.cs:35]: Grenades have a short global cooldown and longer indivodual cooldowns.

File: SWLOR.Game.Server/Feature/UsePerkFeat.cs (hunk 3 above)
Comment (Valo56) [reply SWLOR.Game.Server/Feature/UsePerkFeat.cs:244]: I'm not sure what the breaking point is. What scenario are we concerned about? Can investigate.

File: SWLOR.Game.Server/Feature/StatusEffectDefinition/ForceHealStatusEffectDefinition.cs [hunk 6]
```diff
@@ -23,11 +24,11 @@ public class ForceHealStatusEffectDefinition : IStatusEffectListDefinition
 
         private void ApplyHeal(uint source, uint target, int amount)
         {
-            var wilBonus = GetAbilityModifier(AbilityType.Willpower, source) * 2;
+            var wilBonus = GetAbilityModifier(AbilityType.Willpower, source);
             if (wilBonus < 0)
                 wilBonus = 0;
 
-            amount += wilBonus;
+            amount += (amount / 5 * wilBonus) + Random.D10(wilBonus/2);
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/StatusEffectDefinition/ForceHealStatusEffectDefinition.cs:31]: Just checking you intend this, but this is going to divide your amount by 5 _before_ multiplying by the Will bonus. Might be intentional, just calling it out.

Comment (Valo56) [reply SWLOR.Game.Server/Feature/StatusEffectDefinition/ForceHealStatusEffectDefinition.cs:31]: I'll go over the math later and double check this isn't in error.

File: SWLOR.Game.Server/Service/AbilityService/RecastGroup.cs (hunk 1 above)
Comment (Valo56) [reply SWLOR.Game.Server/Service/AbilityService/RecastGroup.cs:29]: Players don't really see recast timers so I can just change it back to force body and put force body/mind on force body's timer.

Otherwise I'll talk with you later to check on how to handle the migration stuff.

File: SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs [hunk 7]
```diff
@@ -344,23 +344,21 @@ private void ForceBurst()
         }
         private void ForceMind()
         {
-            _builder.Create(PerkCategoryType.ForceLight, PerkType.ForceMind)
+            _builder.Create(PerkCategoryType.ForceUniversal, PerkType.ForceMind)
                 .Name("Force Mind")
 
                 .AddPerkLevel()
-                .Description("Converts 25% of the user's FP into HP.")
+                .Description("PLACEHOLDER.")
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs:351]: Placeholder text here needs to be fixed.

File: SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs [hunk 8]
```diff
@@ -344,23 +344,21 @@ private void ForceBurst()
...
                 .Price(3)
                 .RequirementSkill(SkillType.Force, 20)
                 .RequirementCharacterType(CharacterType.ForceSensitive)
-                .RequirementCannotHavePerk(PerkType.ForceBody)
                 .GrantsFeat(FeatType.ForceMind1)
 
                 .AddPerkLevel()
-                .Description("Converts 50% of the user's FP into HP.")
+                .Description("PLACEHOLDER.")
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs:358]: Same thing here.

File: SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs (hunk 7 above)
Comment (Valo56) [reply SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs:351]: woops

yep

File: SWLOR.Game.Server/Feature/AbilityDefinition/Force/ThrowLightsaberAbilityDefinition.cs [hunk 9]
```diff
@@ -129,9 +131,11 @@ private static void ThrowLightsaber1(AbilityBuilder builder)
             builder.Create(FeatType.ThrowLightsaber1, PerkType.ThrowLightsaber)
                 .Name("Throw Lightsaber I")
                 .Level(1)
-                .HasRecastDelay(RecastGroup.ThrowLightsaber, 30f)
+                .HasRecastDelay(RecastGroup.ThrowLightsaber, 18f)
+                .HasActivationDelay(1.5f)
                 .HasMaxRange(15.0f)
-                .RequirementFP(2)
+                .RequirementFP(1)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Force/ThrowLightsaberAbilityDefinition.cs:137]: Requiring both FP and STM is an interesting choice.

File: SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs [hunk 10]
```diff
@@ -77,35 +77,35 @@ private void ThrowRock()
...
-                .Description("Telekinetically launches a chunk of the environment at the enemy. Deals 10 physical DMG to a single target.")
+                .Description("Telekinetically launches a chunk of the environment at the enemy. Deals 10 + Willpower Mod * 2 physical DMG to a single target.")
                 .Price(1)
                 .RequirementSkill(SkillType.Force, 10)
                 .RequirementCharacterType(CharacterType.ForceSensitive)
                 .GrantsFeat(FeatType.ThrowRock1)
 
                 .AddPerkLevel()
-                .Description("Telekinetically launches a chunk of the environment at the enemy. Deals 15 physical DMG to a single target.")
+                .Description("Telekinetically launches a chunk of the environment at the enemy. Deals 20 + Willpower Mod * 3 physical DMG to a single target.")
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs:87]: I'm really on the fence on giving exact math in the description. I feel it can cause confusion to players who don't care about it and just want to know what to raise to improve the scaling. It's a large part of the reason I chose to stick to a single "DMG" value - simplicity. 

The other thing is it's very error-prone. If you make a math change you also have to remember to come back here and update it.

I think I lean more toward simplifying this by saying something similar to "scales with WIL" or something like that. If we want exact numbers, those can go on the Wiki instead.

Thoughts?

File: SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs [hunk 11]
```diff
@@ -526,23 +522,23 @@ private void Benevolence()
...
+                .Description("Restores 30 + Willpower Mod * 4 + (Willpower Mod*4)d10 health to target. If targeting another, healing is doubled, cost is increased by 10 FP. Willpower Mod/2 stamina is taken (bonuses not lost on insufficient stamina), to restore Willpower Mod/2 Stamina and FP to target and provide Willpower Mod * 2 regen for 4 rounds.")
                 .Price(2)
                 .RequirementSkill(SkillType.Force, 10)
                 .RequirementCharacterType(CharacterType.ForceSensitive)
                 .RequirementCannotHavePerk(PerkType.CreepingTerror)
                 .GrantsFeat(FeatType.Benevolence1)
 
                 .AddPerkLevel()
-                .Description("Restores 80 HP to a single target.")
+                .Description("Restores 60 + Willpower Mod * 4 + (Willpower Mod*4)d10 health to target. If targeting another, healing is doubled, cost is increased by 10 FP. Willpower Mod/2 stamina is taken (bonuses not lost on insufficient stamina), to restore Willpower Mod/2 Stamina and FP to target and provide Willpower Mod * 2 regen for 4 rounds.")
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs:533]: This is a good example of what I meant earlier about simplicity. My eyes glazed over just reading this description. ;)

File: SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs (hunk 10 above)
Comment (Valo56) [reply SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs:87]: Yeah that works for me.

File: SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs [hunk 12]
```diff
@@ -608,23 +604,23 @@ private void CreepingTerror()
...
+                .Description("Attempts to Paralyze a target for 3 seconds with a 12DC will check and inflicts Terror which deals 5 + Willpower Modifier DMG every six seconds for 24 seconds.")
                 .Price(2)
                 .RequirementSkill(SkillType.Force, 20)
                 .RequirementCharacterType(CharacterType.ForceSensitive)
                 .RequirementCannotHavePerk(PerkType.Benevolence)
                 .GrantsFeat(FeatType.CreepingTerror2)
 
                 .AddPerkLevel()
-                .Description("Attempts to Paralyze a target for 2 seconds with a 14DC will check and inflicts Terror which deals 16 DMG every six seconds for 24 seconds.")
+                .Description("Attempts to Paralyze a target for 3 seconds with a 14DC will check and inflicts Terror which deals 10 + Willpower MOdifier * 2 DMG every six seconds for 24 seconds.")
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs:623]: Typo here. MOdifier -> Modifier.

File: SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs [hunk 13]
```diff
@@ -450,23 +448,21 @@ private void ForceLightning()
...
+                .Description("Restores Willpower Mod / 2 Force Points per round.")
                 .Price(3)
                 .RequirementSkill(SkillType.Force, 20)
                 .RequirementCharacterType(CharacterType.ForceSensitive)
-                .RequirementCannotHavePerk(PerkType.ForceMind)
                 .GrantsFeat(FeatType.ForceBody1)
 
                 .AddPerkLevel()
-                .Description("Converts 50% of the user's HP into FP.")
+                .Description("Restores Willpower Mod Force Points per round.")
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/PerkDefinition/ForcePerkDefinition.cs:462]: If we keep the specific math in the description, I would reword this to something like "Restores FP equivalent to your Willpower Mod every tick."

Try to avoid the terminology for "round" because it's not dependent on the round at all. It just happens to tick at roughly the same amount of time as six seconds. And there's a possibility that it could change in the future.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceSparkAbilityDefinition.cs [hunk 14]
```diff
@@ -67,16 +68,23 @@ private void ForceSpark1()
...
                 .HasMaxRange(10f)
                 .IsHostileAbility()
                 .UsesAnimation(Animation.LoopingConjure1)
                 .DisplaysVisualEffectWhenActivating()
                 .HasImpactAction((activator, target, level, location) =>
                 {
-                    Impact(activator, target, 9, 2, 1, Tier1Tag, 8);
+                    var willBonus = GetAbilityModifier(AbilityType.Willpower, activator);
+                    var willDMG = 10 + (willBonus * 2);
+                    if (Stat.GetCurrentFP(activator) == 0)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceSparkAbilityDefinition.cs:82]: Please change this condition to <= 0. It should be equivalent but if there's ever a bug in the future which results in a negative FP amount, this will catch it.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceSparkAbilityDefinition.cs [hunk 15]
```diff
@@ -85,16 +93,25 @@ private void ForceSpark2()
...
+                .HasActivationDelay(2f)
                 .IsCastedAbility()
                 .HasMaxRange(10f)
                 .IsHostileAbility()
                 .UsesAnimation(Animation.LoopingConjure1)
                 .DisplaysVisualEffectWhenActivating()
                 .HasImpactAction((activator, target, level, location) =>
                 {
-                    Impact(activator, target, 14, 4, 2, Tier2Tag, 12);
+                    var willBonus = GetAbilityModifier(AbilityType.Willpower, activator);
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceSparkAbilityDefinition.cs:105]: Similar to other spots - please wrap this functionality into a method and pass in parameters to it. The logic is the same across both tiers, so there's no reason to duplicate the code. This will improve maintainability as changes only need to occur in one place.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceBurstAbilityDefinition.cs [hunk 16]
```diff
@@ -26,53 +27,70 @@ public class ForceBurstAbilityDefinition : IAbilityListDefinition
...
-                attack,
-                dmg, 
-                attackerStat, 
-                defense, 
-                defenderStat, 
-                0);
-            var delay = GetDistanceBetweenLocations(GetLocation(activator), targetLocation) / 18.0f + 0.35f;
-
-            AssignCommand(activator, () =>
+            var creature = GetFirstObjectInShape(Shape.Sphere, RadiusSize.Medium, GetLocation(target), true, ObjectType.Creature);
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceBurstAbilityDefinition.cs:49]: The radius size on this is medium whereas the Force Lightning one is Huge. From a balance perspective should these be the same?

File: SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceBurstAbilityDefinition.cs [hunk 17]
```diff
@@ -26,53 +27,70 @@ public class ForceBurstAbilityDefinition : IAbilityListDefinition
...
+                    });
+                    DelayCommand(delay + 0.2f, () =>
+                    {
+                        ApplyEffectToObject(DurationType.Instant, EffectVisualEffect(VisualEffect.Vfx_Imp_Pulse_Wind), target);
+                    });
+
+                    CombatPoint.AddCombatPoint(activator, creature, SkillType.Force, 3);
+                    Enmity.ModifyEnmity(activator, creature, 250 * level + damage);
+                }
+                creature = GetNextObjectInShape(Shape.Sphere, RadiusSize.Medium, GetLocation(target), true, ObjectType.Creature);
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceBurstAbilityDefinition.cs:92]: This can cause an infinite loop if the if statement above is never met. Please move line 92 outside of this condition.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceBurstAbilityDefinition.cs [hunk 18]
```diff
@@ -26,53 +27,70 @@ public class ForceBurstAbilityDefinition : IAbilityListDefinition
...
-            CombatPoint.AddCombatPoint(activator, target, SkillType.Force, 3);
+                    var dTarget = creature;
+                    AssignCommand(activator, () =>
+                    {
+                        PlaySound("plr_force_blast");
+                        DoFireball(target);
+                    });
+                    DelayCommand(delay, () =>
+                    {
+                    ApplyEffectToObject(DurationType.Instant, EffectDamage(damage), dTarget);
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceBurstAbilityDefinition.cs:75]: These lines within your delay command need to be tabbed over.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Force/DisturbanceAbilityDefinition.cs [hunk 19]
```diff
@@ -67,16 +68,18 @@ private void Disturbance1()
             _builder.Create(FeatType.Disturbance1, PerkType.Disturbance)
                 .Name("Disturbance I")
                 .Level(1)
-                .HasRecastDelay(RecastGroup.Disturbance, 20f)
-                .RequirementFP(1)
+                .HasRecastDelay(RecastGroup.Disturbance, 6f)
+                .HasActivationDelay(2f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Force/DisturbanceAbilityDefinition.cs:72]: There is no FP cost on this. Intentional?

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/IonGrenadeAbilityDefinition.cs [hunk 20]
```diff
@@ -83,7 +84,15 @@ private void IonGrenade1()
...
                     ExplosiveImpact(activator, location, EffectVisualEffect(VisualEffect.Vfx_Fnf_Electric_Explosion), "explosion1", RadiusSize.Large, (target) =>
                     {
-                        Impact(activator, target, 4, -1);
+                        var perBonus = GetAbilityModifier(AbilityType.Perception, activator);
+                        var perDMG = 10 + perBonus;
+                        var targetRace = GetRacialType(target);
+                        if (targetRace == RacialType.Droid || targetRace == RacialType.Cyborg || targetRace == RacialType.Robot)
+                        {
+                            perDMG = 20 + (perBonus * 2);
+                            return;
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/IonGrenadeAbilityDefinition.cs:93]: This return statement doesn't make any sense.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/IonGrenadeAbilityDefinition.cs [hunk 21]
```diff
@@ -93,8 +102,9 @@ private void IonGrenade2()
             _builder.Create(FeatType.IonGrenade2, PerkType.IonGrenade)
                 .Name("Ion Grenade II")
                 .Level(2)
-                .HasRecastDelay(RecastGroup.IonGrenade, 30f)
-                .HasActivationDelay(1f)
+                .HasRecastDelay(RecastGroup.IonGrenade, 24f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/IonGrenadeAbilityDefinition.cs:106]: You can only have one recast group on a single ability. There shouldn't be a general-purpose "Grenades" group anymore. It should be based on the Ion Grenade recast group now. Please remove.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/IonGrenadeAbilityDefinition.cs [hunk 22]
```diff
@@ -105,7 +115,15 @@ private void IonGrenade2()
...
                     ExplosiveImpact(activator, location, EffectVisualEffect(VisualEffect.Vfx_Fnf_Electric_Explosion), "explosion1", RadiusSize.Large, (target) =>
                     {
-                        Impact(activator, target, 8, 10);
+                        var perBonus = GetAbilityModifier(AbilityType.Perception, activator);
+                        var perDMG = 20 + (perBonus * 2);
+                        var targetRace = GetRacialType(target);
+                        if (targetRace == RacialType.Droid || targetRace == RacialType.Cyborg || targetRace == RacialType.Robot)
+                        {
+                            perDMG = 40 + (perBonus * 4);
+                            return;
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/IonGrenadeAbilityDefinition.cs:124]: Same thing here - return statement doesn't make sense.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/IonGrenadeAbilityDefinition.cs [hunk 23]
```diff
@@ -71,8 +71,9 @@ private void IonGrenade1()
             _builder.Create(FeatType.IonGrenade1, PerkType.IonGrenade)
                 .Name("Ion Grenade I")
                 .Level(1)
-                .HasRecastDelay(RecastGroup.IonGrenade, 30f)
-                .HasActivationDelay(1f)
+                .HasRecastDelay(RecastGroup.IonGrenade, 24f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/IonGrenadeAbilityDefinition.cs:75]: Same thing here - only one recast group is allowed and it should be the specific Ion Grenade group.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FragGrenadeAbilityDefinition.cs [hunk 24]
```diff
@@ -66,8 +66,9 @@ private void FragGrenade1()
             _builder.Create(FeatType.FragGrenade1, PerkType.FragGrenade)
                 .Name("Frag Grenade I")
                 .Level(1)
-                .HasRecastDelay(RecastGroup.FragGrenade, 30f)
-                .HasActivationDelay(1f)
+                .HasRecastDelay(RecastGroup.FragGrenade, 12f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FragGrenadeAbilityDefinition.cs:70]: More recast group issues.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FragGrenadeAbilityDefinition.cs [hunk 25]
```diff
@@ -88,8 +91,9 @@ private void FragGrenade2()
             _builder.Create(FeatType.FragGrenade2, PerkType.FragGrenade)
                 .Name("Frag Grenade II")
                 .Level(2)
-                .HasRecastDelay(RecastGroup.FragGrenade, 30f)
-                .HasActivationDelay(1f)
+                .HasRecastDelay(RecastGroup.FragGrenade, 12f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FragGrenadeAbilityDefinition.cs:95]: More recast group issues.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FragGrenadeAbilityDefinition.cs [hunk 26]
```diff
@@ -110,8 +116,9 @@ private void FragGrenade3()
             _builder.Create(FeatType.FragGrenade3, PerkType.FragGrenade)
                 .Name("Frag Grenade III")
                 .Level(3)
-                .HasRecastDelay(RecastGroup.FragGrenade, 30f)
-                .HasActivationDelay(1f)
+                .HasRecastDelay(RecastGroup.FragGrenade, 12f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FragGrenadeAbilityDefinition.cs:120]: More recast group issues.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FlashbangDefinitionAbilityDefinition.cs [hunk 27]
```diff
@@ -38,8 +38,9 @@ private void FlashbangGrenade1()
             _builder.Create(FeatType.FlashbangGrenade1, PerkType.FlashbangGrenade)
                 .Name("Flashbang Grenade I")
                 .Level(1)
-                .HasRecastDelay(RecastGroup.FlashbangGrenade, 30f)
-                .HasActivationDelay(1f)
+                .HasRecastDelay(RecastGroup.FlashbangGrenade, 24f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FlashbangDefinitionAbilityDefinition.cs:42]: More recast group issues.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FlashbangDefinitionAbilityDefinition.cs [hunk 28]
```diff
@@ -60,8 +61,9 @@ private void FlashbangGrenade2()
             _builder.Create(FeatType.FlashbangGrenade2, PerkType.FlashbangGrenade)
                 .Name("Flashbang Grenade II")
                 .Level(2)
-                .HasRecastDelay(RecastGroup.FlashbangGrenade, 30f)
-                .HasActivationDelay(1f)
+                .HasRecastDelay(RecastGroup.FlashbangGrenade, 24f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FlashbangDefinitionAbilityDefinition.cs:65]: More recast group issues.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FlashbangDefinitionAbilityDefinition.cs [hunk 29]
```diff
@@ -82,8 +84,9 @@ private void FlashbangGrenade3()
             _builder.Create(FeatType.FlashbangGrenade3, PerkType.FlashbangGrenade)
                 .Name("Flashbang Grenade III")
                 .Level(3)
-                .HasRecastDelay(RecastGroup.FlashbangGrenade, 30f)
-                .HasActivationDelay(1f)
+                .HasRecastDelay(RecastGroup.FlashbangGrenade, 24f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/FlashbangDefinitionAbilityDefinition.cs:88]: More recast group issues.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/ConcussionGrenadeAbilityDefinition.cs [hunk 30]
```diff
@@ -68,8 +68,9 @@ private void ConcussionGrenade1()
             _builder.Create(FeatType.ConcussionGrenade1, PerkType.ConcussionGrenade)
                 .Name("Concussion Grenade I")
                 .Level(1)
-                .HasRecastDelay(RecastGroup.ConcussionGrenade, 30f)
-                .HasActivationDelay(1f)
+                .HasRecastDelay(RecastGroup.ConcussionGrenade, 24f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/ConcussionGrenadeAbilityDefinition.cs:72]: More recast group issues.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/ConcussionGrenadeAbilityDefinition.cs [hunk 31]
```diff
@@ -92,8 +95,9 @@ private void ConcussionGrenade2()
             _builder.Create(FeatType.ConcussionGrenade2, PerkType.ConcussionGrenade)
                 .Name("Concussion Grenade II")
                 .Level(2)
-                .HasRecastDelay(RecastGroup.ConcussionGrenade, 30f)
-                .HasActivationDelay(1f)
+                .HasRecastDelay(RecastGroup.ConcussionGrenade, 24f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/ConcussionGrenadeAbilityDefinition.cs:99]: More recast group issues.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/AdhesiveGrenadeAbilityDefinition.cs [hunk 32]
```diff
@@ -48,7 +48,8 @@ private void AdhesiveGrenade1()
             _builder.Create(FeatType.AdhesiveGrenade1, PerkType.AdhesiveGrenade)
                 .Name("Adhesive Grenade I")
                 .Level(1)
-                .HasRecastDelay(RecastGroup.AdhesiveGrenade, 30f)
+                .HasRecastDelay(RecastGroup.AdhesiveGrenade, 24f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/AdhesiveGrenadeAbilityDefinition.cs:52]: More recast group issues.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/AdhesiveGrenadeAbilityDefinition.cs [hunk 33]
```diff
@@ -70,7 +71,8 @@ private void AdhesiveGrenade2()
             _builder.Create(FeatType.AdhesiveGrenade2, PerkType.AdhesiveGrenade)
                 .Name("Adhesive Grenade II")
                 .Level(2)
-                .HasRecastDelay(RecastGroup.AdhesiveGrenade, 30f)
+                .HasRecastDelay(RecastGroup.AdhesiveGrenade, 24f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/AdhesiveGrenadeAbilityDefinition.cs:75]: More recast group issues.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Devices/AdhesiveGrenadeAbilityDefinition.cs [hunk 34]
```diff
@@ -92,7 +94,8 @@ private void AdhesiveGrenade3()
             _builder.Create(FeatType.AdhesiveGrenade3, PerkType.AdhesiveGrenade)
                 .Name("Adhesive Grenade III")
                 .Level(3)
-                .HasRecastDelay(RecastGroup.AdhesiveGrenade, 30f)
+                .HasRecastDelay(RecastGroup.AdhesiveGrenade, 24f)
+                .HasRecastDelay(RecastGroup.Grenades, 6f)
```
Comment (zunath) [inline SWLOR.Game.Server/Feature/AbilityDefinition/Devices/AdhesiveGrenadeAbilityDefinition.cs:98]: More recast group issues.

**Review by zunath — CHANGES_REQUESTED** [review]
Requesting changes

File: SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceBurstAbilityDefinition.cs (hunk 16 above)
Comment (Valo56) [reply SWLOR.Game.Server/Feature/AbilityDefinition/Force/ForceBurstAbilityDefinition.cs:49]: The radius of Lightning is huge because it will only target up to a few enemies in that radius. Meanwhile, Burst has a smaller radius but targets everything in that radius.

File: SWLOR.Game.Server/Feature/AbilityDefinition/Force/DisturbanceAbilityDefinition.cs (hunk 19 above)
Comment (Valo56) [reply SWLOR.Game.Server/Feature/AbilityDefinition/Force/DisturbanceAbilityDefinition.cs:72]: Nope. Gotta fix.
